### ISSUE TITLE
Bug: enabled forceGlobalSecureRequests prevents uri schemes besides http/https

### ISSUE THREAD

**Issue by element-code** [codeigniter4/CodeIgniter4#5728]
[HTTP/URI.php#L599](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/URI.php#L599) assumes the only URI schemes are `http` and `https` and therefore ignores other URI schemes when using `forceGlobalSecureRequests`.

This is allread tackled in [Helpers/url_helper.php#L65](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/Helpers/url_helper.php#L65) and  [HTTP/IncomingRequest.php#L411](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L411).
Since this is required in multiple places and therfore error-prone, maybe consider outsourcing this to one tested static method.
Even introduced this differing behaviors in the same commit https://github.com/codeigniter4/CodeIgniter4/commit/89d588fab6dadc904c76859b4a8a86a3e8d11189

```php
d( (string) new URI('ftp://localhost') )
// Wrong Output -> "https://localhost"
// Expected Output -> "ftp://localhost"
```

**Comment by kenjis** [comment]
What's your exact problem?
Why do you use ftp in HTTP/URI?

**Comment by element-code** [comment]
I need to process URIs progrmatically/dynamically so using HTTP/Uri as the OOP abstraction layer only makes sense in my eyes.
In my spcific case this is not limited to ftp, there are many URI schemes involved, even custom application uri schemes.

**Comment by kenjis** [comment]
Thank you for explanation.
Because of language barrier, it is sometimes difficult to me to understand what's saying.

At first I thought the URI class represents URI in HTTP, because the name is `HTTP\URI`,
but it represents all URI because of the user guide.
https://codeigniter.com/user_guide/libraries/uri.html

So this is a bug.

**Comment by kenjis** [comment]
It seems the URI class is very tightly coupled with the CodeIgniter's HTTP processing.
To be honest, I don't recommend to use it for general URI processing...

**Comment by kenjis** [comment]
The part that is tightly coupled seems to be only `__toString()`.

**Comment by kenjis** [comment]
@element-code Fixed and the fix will be included in v4.2.0.

Due to the historical reason,
when the scheme is http(s) and the hostname matches baseURL's hostname,
the URI class may still change the URL.

https://github.com/codeigniter4/CodeIgniter4/blob/200a04eab04262ce79a62b4a776c748f857b73cd/system/HTTP/URI.php#L585
https://github.com/codeigniter4/CodeIgniter4/blob/200a04eab04262ce79a62b4a776c748f857b73cd/system/HTTP/URI.php#L596-L602

**Comment by element-code** [comment]
Thanks for the work and the quick solve @kenjis 👍
